File: guests/java/dev/kaya/guests/Sheet.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

/**
 * The sheet scene from the JVM — guests/rust/sheet.rs, tools/scenes/sheet.steps.
 */
public final class Sheet {
    private static final long TASK = 11;
    private static final long DETAILS = 12;

    public static void app() {
        KayaApp app = new KayaApp();

        KayaApp.Signal<String> status = app.build(tx -> {
            tx.window(0).title("sheet");
            KayaApp.Signal<String> s = tx.signal("closed");
            KayaApp.Signal<String> draft = tx.signal("draft: none");
            tx.mount(tx.column(col -> {
                tx.label(s); // label#0
                tx.button("new task", inner -> openTask(inner, s, draft, false)); // button#0
                tx.button("new task, armed", inner -> openTask(inner, s, draft, true)); // button#1
            }));
            return s;
        });

        if (status == null) throw new IllegalStateException();

        app.dispatchLoop();
    }

    private static void openTask(KayaApp.Tx inner, KayaApp.Signal<String> status,
                                 KayaApp.Signal<String> draft, boolean armed) {
        KayaApp.SheetRef sheet = inner.presentSheet(TASK)
                .title("new task")
                .detent(KayaApp.Detent.MEDIUM)
                .interceptDismiss(armed)
                .onDismissed(tx2 -> tx2.write(status, "dismissed"));
        if (armed) {
            // Nothing has gone; the app keeps the sheet up and says so.
            sheet.onDismissRequested(tx2 -> tx2.write(status, "dismiss requested"));
        }
        KayaApp.Widget body = inner.column(col -> {
            KayaApp.Signal<String> caption = inner.signal("what needs doing?");
            inner.label(caption); // label#1
            inner.entry((tx2, text) -> tx2.write(draft, "draft: " + text)); // entry#0
            inner.label(draft); // label#2
            inner.button("details", tx2 -> { // button#2
                long child = tx2.presentSheetOver(TASK, DETAILS)
                        .title("details")
                        .onDismissed(tx3 -> tx3.write(status, "details dismissed"))
                        .id();
                KayaApp.Widget pane = tx2.column(col2 -> {
                    KayaApp.Signal<String> more = tx2.signal("more about it");
                    tx2.label(more);
                });
                tx2.mountIn(child, pane);
                tx2.write(status, "details open");
            });
            inner.button("done", tx2 -> { // button#3
                // Programmatic: no sheet_dismissed follows, so "done" stays.
                tx2.dismissSheet(TASK);
                tx2.write(status, "done");
            });
        });
        inner.mountIn(sheet.id(), body);
        inner.write(status, "open");
        inner.write(draft, "draft: none");
    }

    private Sheet() {}
}

File: guests/go/sheet/sheet.go
// The sheet scene (tools/scenes/sheet.steps; docs/sheet-plan.md §6): a
// sheet opened from a button and closed through the platform's cancel
// path, the same sheet with the dismiss veto armed, a child sheet chained
// over it, and a programmatic dismiss that echoes nothing.
package sheet

import (
	kaya "dev.kaya/bindings/go"
)

const (
	task    = 11
	details = 12
)

func App() *kaya.App {
	app := kaya.NewApp()

	var status, draft kaya.Signal[string]
	app.Build(func(tx *kaya.Tx) {
		tx.Window(0).Title("sheet")
		status = tx.Signal("closed")
		draft = tx.Signal("draft: none")

		openTask := func(tx *kaya.Tx, armed bool) {
			sheet := tx.PresentSheet(task).
				Title("new task").
				Detent(kaya.DetentMedium).
				InterceptDismiss(armed).
				OnDismissed(func(tx *kaya.Tx) { tx.Write(status, "dismissed") })
			if armed {
				// Nothing has gone; the app keeps the sheet up and says so.
				sheet = sheet.OnDismissRequested(func(tx *kaya.Tx) {
					tx.Write(status, "dismiss requested")
				})
			}
			body := tx.Column(func() {
				caption := tx.Signal("what needs doing?")
				tx.Label(caption)                         // label#1
				tx.Entry(func(tx *kaya.Tx, text string) { // entry#0
					tx.Write(draft, "draft: "+text)
				})
				tx.Label(draft)                          // label#2
				tx.Button("details", func(tx *kaya.Tx) { // button#2
					child := tx.PresentSheetOver(task, details).
						Title("details").
						OnDismissed(func(tx *kaya.Tx) { tx.Write(status, "details dismissed") }).
						Id()
					pane := tx.Column(func() {
						caption := tx.Signal("more about it")
						tx.Label(caption)
					})
					tx.MountIn(child, pane)
					tx.Write(status, "details open")
				})
				tx.Button("done", func(tx *kaya.Tx) { // button#3
					// Programmatic: no sheet_dismissed follows, so "done" stays.
					tx.DismissSheet(task)
					tx.Write(status, "done")
				})
			})
			tx.MountIn(sheet.Id(), body)
			tx.Write(status, "open")
			tx.Write(draft, "draft: none")
		}

		tx.Mount(tx.Column(func() {
			tx.Label(status)                                                       // label#0
			tx.Button("new task", func(tx *kaya.Tx) { openTask(tx, false) })       // button#0
			tx.Button("new task, armed", func(tx *kaya.Tx) { openTask(tx, true) }) // button#1
		}))
	})

	return app
}
